### ISSUE TITLE
[Feature] Adding `conditions` to `actions`?

### ISSUE THREAD

**Issue by sebastianfeldmann** [CaptainHookPhp/captainhook#27]
I wanted to write an `Action` that executes `composer install` after a merge or a checkout if `composer.json` or `composer.lock` where changed.

First I created an `Action` called `...\Composer\Action\TriggerInstall`, then I realized I could write it in a more generic way. `...\Change\Action\CmdTrigger` where you could specify the list of files to watch and a cli command to execute.

But being limited to calling cli commands didn't feel right. So in order to allow PHP "commands" as well I could write something like `...\Change\Action\PHPTrigger`.
But passing options to this "command" would create an action config in an action config and that feels even more strange.

To describe the problem in a very abstract way I want to to execute a certain action only under specific circumstances. So I thought about adding `conditions` to an `action` configuration.

This could look like this.

```json
    "post-checkout": {
        "enabled": true,
        "actions": [
            {
                "action": "composer install",
                "options": [],
                "conditions": [
                    {
                        "class": "\\CaptainHook\\App\\Hook\\Condition\\FileChange",
                        "args": {
                            "files": [
                                "composer.json",
                                "composer.lock"
                            ]
                        }
                    }
                ]
            }
        ]
    }
```

This would blow up the configuration quite a bit but would be a very flexible solution. People could write their own `conditions` by implementing a simple one method interface.

Any feedback would be appreciated.

**Comment by sebastianfeldmann** [comment]
@heiglandreas @shochdoerfer what do you think

**Comment by heiglandreas** [comment]
I like the idea as it is as general as possible. The addition to the config is minimal so that would not be a blocker for me. and as it is an optional addition it should be BC as well. So from my side: :+1:

**Comment by shochdoerfer** [comment]
Interesting idea. What about not limiting conditions to be php classes but also also external tools? Similar to what's possible with actions?

Since the configuration option is named `conditions` would you also want to support `and`, `or`, `not` operators or just stick to something hardcoded?

**Comment by sebastianfeldmann** [comment]
Ok let's imagine a configuration that allows that

```json    "post-checkout": {
        "enabled": true,
        "actions": [
            {
                "action": "composer install",
                "options": [],
                "comment": "Conditions are combined with OR, inner conditions with AND",
                "conditions": [
                    {
                        "inner": [
                            {
                                "condition": "\\CaptainHook\\App\\Hook\\Condition\\FileChange",
                                "options": {
                                    "files": [
                                        "composer.json",
                                        "composer.lock"
                                    ]
                                },
                                "mode": "NOT"
                            },
                            {
                                "condition": "git diff-tree -r -no-revision HEAD@{1} HEAD | lc",
                            }
                        ]
                    }
                ]
            }
        ]
    }
```

I'm not sure I like the complexity of the configuration in this scenario. I think I would prefer to write a custom `condition` for something like this, especially if you have to duplicate the configuration for example in `post-merge` and `post-checkout`. And, to "really" support `OR` and `AND` the configuration has to become even more complex.

I'm struggling with the decision if binaries and static php method calls should be supported for conditions as well.
But I'm leaning more towards "not for now, but build it in a way they can be added without BC breaks at a later point in time".

**Comment by shochdoerfer** [comment]
I agree with you and since there's no way to express something similar for the actions, it makes sense to stick with a similar approach. I am with you that for complex logic writing a php class makes more sense.

I am fine with "classes for conditions for now, maybe commands later".

**Comment by heiglandreas** [comment]
I'd probably add an `AndCondition`, `OrCondition` and `NotCondition` implementing the `ConditionInterface` that all accept one or more `ConditionInterface`s to delegate the request to. And whether the final `ConditionInterface` then calls a PHP function or a shellscript depends on the implementation...

**Comment by shochdoerfer** [comment]
@heiglandreas but that means you somehow need to define the dependency hierarchy in the JSON structure, not sure how suitable that is.

**Comment by sebastianfeldmann** [comment]
For now I’m planning to combine conditions by `and`.
So if you specify multiple conditions, they all have to apply.
I think that is the most intuitive solution.

PS: Looks like CLI conditions will be supported from the get go but no promises

**Comment by sebastianfeldmann** [comment]
A first version of this is now implemented and included in version 4.2.0.
To get an impression how it works have a look at the [documentation](https://captainhookphp.github.io/captainhook/configure.html#conditions)
